theseus/embodied/kinematics/kinematics_model.py

```diff
@@ -0,0 +1,112 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import abc
+from typing import Dict, Optional, Union
+
+import differentiable_robot_model as drm
+import torch
+
+from theseus import Variable
+from theseus.geometry import SE3, LieGroup, Point2, Vector
+
+RobotModelInput = Union[torch.Tensor, Vector]
+
+
+class KinematicsModel(abc.ABC):
+    def __init__(self):
+        pass
+
+    @abc.abstractmethod
+    def forward_kinematics(self, robot_pose: RobotModelInput) -> Dict[str, LieGroup]:
+        pass
+
+    @abc.abstractmethod
+    def dim(self) -> int:
+        pass
+
+
+class IdentityModel(KinematicsModel):
+    def __init__(self):
+        super().__init__()
+
+    def forward_kinematics(self, robot_pose: RobotModelInput) -> Dict[str, LieGroup]:
+        if isinstance(robot_pose, Point2) or isinstance(robot_pose, Vector):
+            assert robot_pose.dof() == 2
+            return {"state": robot_pose}
+        raise NotImplementedError(
+            f"IdentityModel not implemented for pose with type {type(robot_pose)}."
+        )
+
+    def dim(self) -> int:
+        return 1
+
+
+class UrdfRobotModel(KinematicsModel):
+    def __init__(self, urdf_path: str):
+        self.drm_model = drm.DifferentiableRobotModel(urdf_path)
+
+    def _postprocess_quaternion(self, quat):
+        # Convert quaternion convention (DRM uses xyzw, Theseus uses wxyz)
+        quat_converted = torch.cat([quat[..., 3:], quat[..., :3]], dim=-1)
+
+        # Normalize quaternions
+        quat_normalized = quat_converted / torch.linalg.norm(
+            quat_converted, dim=-1, keepdim=True
+        )
+
+        return quat_normalized
+
+    def forward_kinematics(
+        self,
+        joint_states: RobotModelInput,
+        jacobians: Dict[str, Optional[torch.Tensor]] = {},
```

Comment:
The `Optional` on tensor means we just pass a string, but then where is the computed jacobian stored?

We might want to add `Optional[Dict[...` at the start so we can then avoid computing jacobians when not needed and be more efficient.

Reply:
My understanding of the example Luis provided is that we'll be passing in a dictionary of link names with values being None, and the `forward_kinematics` function will then populate those values with the actual jacobian. 
Note that I also plan to update DRM to compute jacobians for all links in a more efficient way, but I feel like that is lower pri because 1) not all usecases of forward kinematics uses jacobian computation; 2) I imagine it will be a rare case where users will need access to the Jacobians of all links.

Reply:
@luisenp Can you confirm/clarify what the exact interface should be? Thanks!

Reply:
That makes sense. In other places [like here](https://github.com/facebookresearch/theseus/pull/84/files#diff-0579a4f9834772c89a164d4e1f36069e7eab1168c2ce477c109be11c31c024cdR86-R98) we optionally pass a tensor for a jacobian to be computed. Luis can confirm if we want to follow a similar API here.

Reply:
On a closer look, I think the current version is similar to api elsewhere the difference is the Dict.

Reply:
What I had in mind was `jacobians: Optional[Dict[str, torch.Tensor]] = None`, and the function will only populate jacobians if the dict is not `None` and only for any keys that are present in the dict. 

PS: BTW, it's good to avoid passing mutable objects as defaults (e.g., `{}`) because these defaults are actually associated to the class, not the object.

Reply:
Good catch on passing mutable objects, this can actually introduce some hard to find bugs.

Reply:
Should the values also be `Optional` since we might be passing in {ee_name: None}?
We could also just do `jacobians: Optional[Dict[str, Any]] = None` since the values are going to be overwritten anyways.

Reply:
Ah, good point! It feels a bit strange to me to pass `None` values. Would it be better to do the following?

```python
def forward_kinematics(
        self,
        joint_states: RobotModelInput,
        jacobians: Optional[Dict[str, torch.Tensor]] = None,
        jacobians_links: Optional[List[str]] = None, 
)
```

If the `jacobians_links` list is not passed, then jacobians for all the links are computed. Any thoughts?

Reply:
You would still need to then make a Dict with all the link names for jacobians, which seems redundant.

`Optional[Dict[str, Optional[torch.Tensor]]] = None` seems like the cleanest option to me. Then a user can also pass `jacobians = dict.fromkeys(['link1', 'link2'])` with `None` in the values.